src/WaveFunction.cpp

```diff
@@ -0,0 +1,256 @@
+    SolverControl solver_control(1000, 1e-8 * system_rhs.l2_norm());
+    SolverCG<Vector<double>> cg(solver_control);
+    cg.solve(matrix_v, solution_v, system_rhs, PreconditionIdentity());
+  }
+  //end 
+//================================================================================================================================
+  //start function for generating output
+  template <int dim>
+  void WaveEquation<dim>::output_results() const
+  {
+    // create VTK output file with solution data
+    DataOut<dim> data_out;
+    data_out.attach_dof_handler(dof_handler);
+    data_out.add_data_vector(solution_u, "U");
+    data_out.add_data_vector(solution_v, "V");
+    data_out.build_patches();
+
+    // create the output filename based on the current timestep
+    const std::string filename =
+      "solution-" + Utilities::int_to_string(timestep_number, 3) + ".vtu";
+    const std::string full_filename = "output/" + filename;
+
+    // set compression level for faster writing
+    DataOutBase::VtkFlags vtk_flags;
+    vtk_flags.compression_level = DataOutBase::CompressionLevel::best_speed;
+    data_out.set_flags(vtk_flags);
+
+    // write the VTU file
+    std::ofstream output(full_filename);
+    data_out.write_vtu(output);
+
+    // update the PVD file for time-series visualization in ParaView
+    static std::ofstream pvd_file("output/solution.pvd", std::ios::app);
+    static bool first_time = true;
+
+    if (first_time)
+    {
+      pvd_file << "<?xml version=\"1.0\"?>\n"
+               << "<VTKFile type=\"Collection\" version=\"0.1\" byte_order=\"LittleEndian\">\n"
+               << "  <Collection>\n";
+      first_time = false;
+    }
+
+    // write a reference to the new .vtu file with its associated timestep
+    pvd_file << "    <DataSet timestep=\"" << time
+             << "\" group=\"\" part=\"0\" file=\"" << filename << "\"/>\n";
+
+    // close the PVD file at the end of the simulation
+    if (time + time_step > 5.0)
+    {
+      pvd_file << "  </Collection>\n"
+               << "</VTKFile>\n";
+      pvd_file.close();
+    }
+  }
+  //end 
+//================================================================================================================================
+  //start run function 
+  template <int dim>
+  void WaveEquation<dim>::run()
+  {
+    // initialize system matrices and vectors
+    setup_system();
+
+    // project initial conditions into the finite element space
+    VectorTools::project(dof_handler, constraints, QGauss<dim>(fe.degree + 1),
+                         InitialValuesU<dim>(), old_solution_u);
+    VectorTools::project(dof_handler, constraints, QGauss<dim>(fe.degree + 1),
+                         InitialValuesV<dim>(), old_solution_v);
+
+    Vector<double> tmp(solution_u.size());
+    Vector<double> forcing_terms(solution_u.size());
+
+    // time-stepping loop from t = 0 to t = 5
+    for (; time <= 5; time += time_step, ++timestep_number)
```

Comment:
This literal `5` is another "magic constant": this is a very bad practice, you should at the very least declare a `const` or `constexpr` variable somewhere.